drivers/spi/spi_mcux_lpspi.c

```diff
@@ -580,14 +617,152 @@ static int spi_mcux_init(const struct device *dev)
 	return 0;
 }
 
+#ifdef CONFIG_SPI_RTIO
+static inline k_spinlock_key_t spi_spin_lock(const struct device *dev)
+{
+	struct spi_mcux_data *data = dev->data;
+
+	return k_spin_lock(&data->lock);
+}
+
+static inline void spi_spin_unlock(const struct device *dev, k_spinlock_key_t key)
+{
+	struct spi_mcux_data *data = dev->data;
+
+	k_spin_unlock(&data->lock, key);
+}
+
+
+static void spi_mcux_iodev_complete(const struct device *dev, int status);
+static void spi_mcux_iodev_next(const struct device *dev, bool completion);
+
+static void spi_mcux_iodev_start(const struct device *dev)
+{
+	struct spi_mcux_data *data = dev->data;
+	struct rtio_sqe *sqe = &data->txn_curr->sqe;
+	struct spi_dt_spec *dt_spec = sqe->iodev->data;
+	struct spi_buf tx_buffer = {};
+	struct spi_buf rx_buffer = {};
+	struct spi_buf_set tx = {};
+	struct spi_buf_set rx = {};
+	int ret = 0;
+
+	switch (sqe->op) {
+	case RTIO_OP_TX:
+	case RTIO_OP_RX:
+		{
+			tx_buffer.buf = 0;
+			tx_buffer.len = sqe->buf_len;
+			rx_buffer.buf = sqe->buf;
+			rx_buffer.len = sqe->buf_len;
+
+			} break;
+	case RTIO_OP_TXRX:
+		{
+			tx_buffer.buf = sqe->tx_buf;
+			tx_buffer.len = sqe->txrx_buf_len;
+			rx_buffer.buf = sqe->rx_buf;
+			rx_buffer.len = sqe->txrx_buf_len;
+		} break;
+	default:
+		{
+			LOG_ERR("Invalid op code %d for submission %p\n", sqe->op, (void *)sqe);
+			struct rtio_iodev_sqe *txn_head = data->txn_head;
+
+			spi_mcux_iodev_next(dev, true);
+			rtio_iodev_sqe_err(txn_head, -EINVAL);
+			ret = 0;
+		}
+	}
+
+	tx.buffers = &tx_buffer;
+	tx.count = 1;
+	rx.buffers = &rx_buffer;
+	rx.count = 1;
+
+	transceive(dev, &dt_spec->config, &tx, &rx, 0, 0, 0);
+	if (ret == 0) {
```

Comment:
when would this not be 0? It seems like this implies a blocking flow at all times.

What should happen is the transceive above should be asynchronously done and upon completion the next transceive started.

Or maybe better is the entire transaction is loaded into a hardware queue and started asynchronously, then upon completion things are marked as such.

Reply:
yes, driver should enable dma if using rtio and call transceive_dma here instead of regular blocking transceive

Reply:
I’m working on adding polling support as well, so if for small transfers it makes more sense to have a tight poll loop, rtio will be able to support that. It should fix the Sam spi issue as well.